torch/csrc/jit/codegen/cuda/lower_mem_index.cpp

```diff
@@ -0,0 +1,1226 @@
+
+      if (!contiguity_id_set.count(first_domain) ||
+          !contiguity_id_set.count(second_domain)) {
+        // Only contiguous pairs of iterdomains are considered.
+        continue;
+      }
+
+      auto first_domain_use_it = id_to_use_map.find(first_domain);
+      TORCH_INTERNAL_ASSERT(first_domain_use_it != id_to_use_map.end());
+
+      auto merge = dynamic_cast<Merge*>(first_domain_use_it->second);
+
+      // See if this is contiguous merge
+      if (merge && merge->inner() == second_domain) {
+        update = true;
+        auto merged_id = merge->out();
+        // Insert the merged id and contiguity
+        domain_list.insert(domain_it, merged_id);
+        contiguity_id_set.erase(first_domain);
+        contiguity_id_set.erase(second_domain);
+        contiguity_id_set.insert(merged_id);
+
+        // Erase the merged id.
+        domain_list.erase(domain_it, std::next(second_domain_it));
+
+        // Start over for simplicity, actually
+        //  just decrement the domain it by one would be enough.
+        break;
+      }
+      // Didn't find a merged domain candidate, just
+      //  increment
+      domain_it++;
+    }
+  }
+
+  std::vector<IterDomain*> result_domain{
+      domain_list.begin(), domain_list.end()};
+  return std::make_pair(result_domain, contiguity_id_set);
+}
+
+//! Utility function that returns a map from each iterdomain from a tensordomain
+//!  to its consumer, assuming that on this tensor domain each iterdomain has
+//!  a unique iterdomain expression that consumes it.
+std::unordered_map<IterDomain*, Expr*> getIdToUseMap(TensorDomain* td) {
+  std::vector<Val*> all_id_vals{td->domain().begin(), td->domain().end()};
+  auto all_inputs = InputsOf::outputs(td->fusion(), all_id_vals);
+  std::unordered_map<IterDomain*, Expr*> id_to_use_map;
+
+  auto all_exprs = DependencyCheck::getAllExprsBetween(
+      {all_inputs.begin(), all_inputs.end()}, all_id_vals);
+
+  for (auto expr : all_exprs) {
+    for (auto id : ir_utils::filterByType<IterDomain>(expr->inputs())) {
+      id_to_use_map[id] = expr;
+    }
+  }
+
+  return id_to_use_map;
+}
+
+//! Given a pair of data_tv and reference_tv, see [Note on data tensor and
+//! reference tensor]:
+//!  Returns the set of iterdomains resulting from all the contiguous merges on
+//!  the reference_tv's tensor domain.
+//! The reason for needing both data_tv and reference_tv is that contiguity is
+//! defined on data_tv
+//!  while the actual index math is defined on reference tv.
+//! TODO: unify with contiguity pass
+std::unordered_set<IterDomain*> getInitialContiguousRootIdsOnReferenceTv(
+    const TensorView* data_tv,
+    const TensorView* reference_tv) {
+  std::pair<std::vector<IterDomain*>, std::unordered_set<IterDomain*>>
+      contig_merge_front;
+
+  // Compute the merge frontier for contiguity info
+  if (data_tv == reference_tv) {
+    contig_merge_front = getContigMergeFrontier(
+        reference_tv->getRootDomain(),
```

Comment:
Should this be rfactor domain instead? Same comment for below root domain usage.